File: JavaFundamentals/JavaAdvanced/LinearDataStructures_Ex/SimpleTextEditor.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;

public class SimpleTextEditor {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(reader.readLine());
        ArrayDeque<String> stackTmp = new ArrayDeque<>();
        String text = "";

        for (int i = 0; i < N; i++) {
            String[] command = reader.readLine().split(" ");
            int index = Integer.parseInt(command[0]);
            switch (index){
                case 1:
                    stackTmp.push(text);
                    text += command[1];
                    break;
                case 2:
                    stackTmp.push(text);
                    int n = text.length();
                    text = text.substring(0, n - Integer.parseInt(command[1]));
                    break;
                case 3:
                    int showIndex = Integer.parseInt(command[1]) - 1;
                    System.out.println(text.charAt(showIndex));
                    break;
                case 4:
                    text = stackTmp.pop();
                    break;
            }
        }
    }
}

File: CSharpFundamentals/CSharpAdvanced/StacksAndQueues_Ex/SimpleTextEditor.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace SimpleTextEditor
{
    class SimpleTextEditor
    {
        static void Main(string[] args)
        {
            int N = int.Parse(Console.ReadLine());
            Stack<string> stackTmp = new Stack<string>();
            StringBuilder text = new StringBuilder();
            for (int i = 0; i < N; i++)
            {
                string[] command = Console.ReadLine().Split(' ', StringSplitOptions.RemoveEmptyEntries);
                int index = int.Parse(command[0]);
                switch (index)
                {
                    case 1:
                        stackTmp.Push(text.ToString());
                        text.Append(command[1]);
                        break;
                    case 2:
                        stackTmp.Push(text.ToString());
                        int count = int.Parse(command[1]);
                        int test = text.Length;
                        text.Remove(Math.Max(0, text.Length - count), Math.Min(text.Length, count));
                        break;
                    case 3:
                        int showIndex = int.Parse(command[1]);
                        Console.WriteLine(text.ToString()[showIndex - 1]);
                        break;
                    case 4:
                        text.Clear();
                        text.Append(stackTmp.Pop());
                        break;
                    default:
                        break;                            
                }
            }
        }
    }
}
